src/app/clusters/window-covering-server/window-covering-server.h

```diff
@@ -0,0 +1,195 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/common/gen/attribute-id.h>
+#include <app/common/gen/enums.h>
+#include <app/util/af-enums.h>
+
+class WindowCover
+{
+public:
+    enum class Feature
+    {
+        Lift          = 0x01,
+        Tilt          = 0x02,
+        PositionAware = 0x04
+    };
+
+    struct Mode
+    {
+        uint8_t motorDirReversed : 1; // bit 0
+        uint8_t calibrationMode : 1;  // bit 1
+        uint8_t maintenanceMode : 1;  // bit 2
+        uint8_t ledDisplay : 1;       // bit 3
+    };
+    static_assert(sizeof(Mode) == sizeof(uint8_t), "Mode Size is not correct");
+
+    struct ConfigStatus
+    {
+        uint8_t operational : 1;             // bit 0 M
+        uint8_t online : 1;                  // bit 1 M
+        uint8_t liftIsReversed : 1;          // bit 2 LF
+        uint8_t liftIsPA : 1;                // bit 3 LF & PA
+        uint8_t tiltIsPA : 1;                // bit 4 TL & PA
+        uint8_t liftIsEncoderControlled : 1; // bit 5 LF & PA
+        uint8_t tiltIsEncoderControlled : 1; // bit 6 LF & PA
+    };
+    static_assert(sizeof(ConfigStatus) == sizeof(uint8_t), "ConfigStatus Size is not correct");
+
+    // Match directly with OperationalStatus 2 bits Fields
+    enum class OperationalState : uint8_t
+    {
+        Stall             = 0x00, // currently not moving
+        MovingUpOrOpen    = 0x01, // is currently opening
+        MovingDownOrClose = 0x02, // is currently closing
+        Reserved          = 0x03, // dont use
+    };
+    static_assert(sizeof(OperationalState) == sizeof(uint8_t), "OperationalState Size is not correct");
+
+    struct OperationalStatus
+    {
+        OperationalState global : 2; // bit 0-1 M
+        OperationalState lift : 2;   // bit 2-3 LF
+        OperationalState tilt : 2;   // bit 4-5 TL
+    };
+    static_assert(sizeof(OperationalStatus) == sizeof(uint8_t), "OperationalStatus Size is not correct");
+
+    struct SafetyStatus
+    {
+        uint16_t remoteLockout : 1;       // bit 0
+        uint16_t tamperDetection : 1;     // bit 1
+        uint16_t failedCommunication : 1; // bit 2
+        uint16_t positionFailure : 1;     // bit 3
+        uint16_t thermalProtection : 1;   // bit 4
+        uint16_t obstacleDetected : 1;    // bit 5
+        uint16_t powerIssue : 1;          // bit 6
+        uint16_t stopInput : 1;           // bit 7
+        uint16_t motorJammed : 1;         // bit 8
+        uint16_t hardwareFailure : 1;     // bit 9
+        uint16_t manualOperation : 1;     // bit 10
+    };
```

Comment:
I'm not sure if this is safe to be used on little and big endian platforms?

Reply:
indeed this is good catch , this won't follow the spec compliance on a big endian machine

    struct message {
        uint16      word;
        struct {
        #ifdef LITTLE_ENDIAN
       uint16_t remoteLockout : 1;       // bit 0
        uint16_t tamperDetection : 1;     // bit 1
        uint16_t failedCommunication : 1; // bit 2
        uint16_t positionFailure : 1;     // bit 3
        uint16_t thermalProtection : 1;   // bit 4
        uint16_t obstacleDetected : 1;    // bit 5
        uint16_t powerIssue : 1;          // bit 6
        uint16_t stopInput : 1;           // bit 7
        uint16_t motorJammed : 1;         // bit 8
        uint16_t hardwareFailure : 1;     // bit 9
        uint16_t manualOperation : 1;     // bit 10
        #else
# all the same in reverse order
        #endif
        } field;
    };

Reply:
This apply to all the bitfields declared 
I heard some people are building on x86 and the SDK had already failure due to endianess 
This must to be fixed